src/workerd/io/worker-fs.h

```diff
@@ -0,0 +1,638 @@
+};
+
+// A directory in the virtual file system. If the directory is read-only,
+// then the mutation methods will throw an exception.
+class Directory: public kj::Refcounted, public kj::EnableAddRefToThis<Directory> {
+ public:
+  // Returns the metadata for this node.
+  virtual kj::Maybe<Stat> stat(jsg::Lock& js, kj::PathPtr ptr) = 0;
+  Stat stat(jsg::Lock& js) {
+    return KJ_ASSERT_NONNULL(stat(js, nullptr));
+  }
+
+  // Return the number of entries in this directory. If typeFilter is provided,
+  // only entries matching the given type will be counted.
+  virtual size_t count(jsg::Lock& js, kj::Maybe<FsType> typeFilter = kj::none) = 0;
+
+  using Item = kj::OneOf<kj::Rc<File>, kj::Rc<Directory>, kj::Rc<SymbolicLink>>;
+  using Entry = kj::HashMap<kj::String, Item>::Entry;
+  virtual Entry* begin() = 0;
+  virtual Entry* end() = 0;
+  virtual const Entry* begin() const = 0;
+  virtual const Entry* end() const = 0;
+
+  struct OpenOptions {
+    // When set, if the path does not exist, we will attempt to create the
+    // node as the specified type. Type must be one of either FILE or DIRECTORY.
+    // Specifying SYMBOLICLINK as the type will throw an exception.
+    kj::Maybe<FsType> createAs;
+
+    // If the path points to a symbolic link, by default the link will be followed
+    // such that if the link points to a valid node, that node will be returned.
+    // If followLinks is false, however, the symbolic link itself will be returned.
+    bool followLinks = true;
+  };
+
+  // Tries opening the file or directory at the given path. If the node does
+  // not exist, and createAs is provided specifying a create mode, the node
+  // will be created as the specified type; otherwise kj::none is returned
+  // if the node does not exist. If the directory is read only and create is
+  // specified, an exception will be thrown.
+  // The path must be relative to the this directory.
+  // Note that when creating files using this method, the newly created file
+  // will have a size of 0 bytes. The file will be writable but each write
+  // could end up causing a new allocation. It the intent is to perform
+  // multiple writes, it is recommended to resize the file to the expected
+  // size up front or to use the newWritable method to create a file with the
+  // expected size and add it into the target directory.
+  //
+  // Note that tryOpen doubles as both an existence check and a stat operation.
+  // When the createAs parameter is not specified, the method will return
+  // kj::none if the file or directory does not exist.
+  //
+  // If the path identifies a symbolic link, the symbolic link will be resolved
+  // and the target, if it exists, will be returned. If the target does not exist,
+  // kj::none will be returned.
+  virtual kj::Maybe<kj::OneOf<kj::Rc<File>, kj::Rc<Directory>, kj::Rc<SymbolicLink>>> tryOpen(
+      jsg::Lock& js, kj::PathPtr path, kj::Maybe<OpenOptions> options = kj::none) = 0;
```

Comment:
Nit:
The following is cleaner and conveys the purpose at the API level without having to go into the implementation.
It also simplifies the implementation a tiny tiny bit
```suggestion
      jsg::Lock& js, kj::PathPtr path, OpenOptions options = {}) = 0;
```